### PR TITLE
Meta folder

### PR DESCRIPTION
**Pull request by nstarman**
There's a lot of different ideas happening in `astropy/utils/metadata.py`: new exceptions and warnings, metadata merging mechanics, a metadata attribute, and attributes stored in metadata. This PR separates these concerns into different files, collecting them in a new module -- `astropy/utils/metadata`. ~There is no change in public API.~
There is change in semi-public API.

Plz squash & merge

### CODE REVIEW

**Comment by nstarman** [comment]
I'm fine to break up this PR into a 2: re-org and improvements. Re the file names, any suggestion?

**Comment by mhvk** [comment]
Yes, I think it should be broken up; for the filenames, I think we should follow precedent and not use underscores. They can be added later if APE 22 is accepted.

**Comment by nstarman** [comment]
SGTM.

**Comment by nstarman** [comment]
Ok. #15237 is merged. This is ready!

**Comment by pllim** [comment]
If there is some change, then need a change log. Thanks!

**Comment by nstarman** [comment]
Changelog added!

**Comment by nstarman** [comment]
Ping @mhvk. Hopefully an easy review now!

File: astropy/utils/metadata/core.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Classes for handling metadata."""
+
+import inspect
+from collections import OrderedDict
+from collections.abc import Mapping
+from copy import deepcopy
+
+__all__ = []
```
Comment (mhvk) [inline astropy/utils/metadata/core.py:9]: Why not follow precedent in `angles` and `representations` and define here what is public at the top level? For better or worse, that is the current standard...

File: astropy/utils/metadata/__init__.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""This module contains helper functions and classes for handling metadata."""
+
+from .core import MetaAttribute, MetaData
```
Comment (mhvk) [inline astropy/utils/metadata/__init__.py:4]: Maybe follow current standard and `import *` from those (defining `__all__`)

File: docs/changes/utils/15166.api.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+The structure of ``utils.meta`` has been refactored, but all the available
```
Comment (mhvk) [inline docs/changes/utils/15166.api.rst:1]: `utils.metadata`

**Review by mhvk — COMMENTED** [review]
We could avoid changes to imports from `test_meta` (and thus the need for possible similar changes outside astropy) if we kept the test files in the same location. Unlike for `astropy.coordinates.angles`, I don't have a strong opinion on what is better. `utils.iers` and `utils.masked` have their tests inside and this is logical; that suggests to do what you have done for `utils.metadata`. On the other hand, for `ies` and (to a less extend) for `masked` there is a case that they don't belong in `utils` but should be on their own, while `metadata` is exactly in the right place. And clearly there are imports from the tests...  Then again, `utils.xml` does have its own tests (though that has C code). Overall, what you have is probably best, but thought I would bring it up anyway.

Otherwise, my only remaining comment is about `__all__` in lower-level files. Let's stick to current precedent for those, for now at least, i.e., fill them with those items that should be imported in `__init__.py`.

File: astropy/utils/metadata/core.py (hunk 1 above)
Comment (nstarman) [reply astropy/utils/metadata/core.py]: Done.

File: docs/changes/utils/15166.api.rst (hunk 3 above)
Comment (nstarman) [reply docs/changes/utils/15166.api.rst]: Done.

File: astropy/utils/metadata/__init__.py (hunk 2 above)
Comment (nstarman) [reply astropy/utils/metadata/__init__.py]: Done.

**Review by mhvk — APPROVED** [review]
Looks all OK to me now! Thanks, @nstarman

**Pull request merged**